File: src/Controller.cs
using BeetleX.Buffers;
using BeetleX.FastHttpApi;
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;

namespace TcpBanchmarks
{
    [BeetleX.FastHttpApi.Controller]
    public class Banchmark
    {
        public void AddServer(string host)
        {
            Uri uri = new Uri(host);
            Service.Default.AddServer(uri.Host, uri.Port);
            Service.Default.Save();
        }

        public void DelServer(string host)
        {
            Uri uri = new Uri(host);
            Service.Default.RemoveServer(uri.Host, uri.Port);
            Service.Default.Save();
        }

        public object ListServer()
        {
            return Service.Default.ListServer();
        }

        public object NewMessage()
        {
            return new Message();
        }

        [Post]
        public void SaveMessage(Message msg)
        {
            Service.Default.SaveMessage(msg);
            Service.Default.Save();
        }

        public void DeleteMessage(string id)
        {
            Service.Default.DeleteMessage(id);
            Service.Default.Save();
        }

        public void DeleteCategoryMessages(string category)
        {
            Service.Default.DeleteCategoryMessages(category);
            Service.Default.Save();
        }

        public object ListMessages(string category)
        {
            return Service.Default.ListMessages(category);
        }

        public object Test(Message message, string host)
        {
            Uri uri = new Uri(host);
            var client = BeetleX.SocketFactory.CreateClient<BeetleX.Clients.TcpClient>(uri.Host, uri.Port);
            using (client)
            {
                byte[] data = message.ToBytes();
                client.TimeOut = 5000;
                client.Send(s => s.Write(data, 0, data.Length));
                var stream = client.Receive();
                data = new byte[stream.Length];
                stream.Read(data, 0, data.Length);
                var hex = Service.Default.BytesToHex(data);
                var utf8 = Encoding.UTF8.GetString(data);
                return new { hex, utf8 };
            }
        }

        public object Download()
        {
            var items = Service.Default.GetStore();
            string text = Newtonsoft.Json.JsonConvert.SerializeObject(items);
            return new DownLoad(text);
        }

        public object Upload(string name, bool completed, string data)
        {
            byte[] array = Convert.FromBase64String(data);
            using (System.IO.Stream writer = System.IO.File.Open(name, System.IO.FileMode.Append))
            {
                writer.Write(array, 0, array.Length);
                writer.Flush();
            }
            if (completed)
            {
                try
                {
                    using (System.IO.StreamReader reader = new System.IO.StreamReader(name))
                    {
                        string value = reader.ReadToEnd();
                        Store setting = Newtonsoft.Json.JsonConvert.DeserializeObject<Store>(value);
                        Service.Default.SetStore(setting);
                    }
                }
                finally
                {
                    System.IO.File.Delete(name);
                }
            }
            return true;
        }

        public object ListMessageCategories()
        {
            return Service.Default.ListMessageCategores();
        }

        private Loadruner mLoadruner = new Loadruner();

        public object GetRunerDetail()
        {
            return mLoadruner?.Detail;
        }

        public object GetFileID()
        {
            return Guid.NewGuid().ToString("N");
        }

        public async Task Stop()
        {

            await Task.Run(() => mLoadruner.Stop());

        }
        [Post]
        public async Task Run(Server server, string[] ipaddress, List<Message> messages, Setting setting)
        {
            if (mLoadruner.Status != RunStatus.Runing && mLoadruner.Status != RunStatus.Init)
            {
                mLoadruner = new Loadruner();
                mLoadruner.Server = server;
                mLoadruner.IPAddress = ipaddress;
                mLoadruner.Setting = setting;
                mLoadruner.Messages = messages;
                await Task.Run(() => mLoadruner.Run());
            }
        }

        public object GetLocalIPAddress()
        {
            var host = Dns.GetHostEntry(Dns.GetHostName());
            var ipaddress = new List<object>();
            ipaddress.Add(new { IP = "127.0.0.1" });
            foreach (var ip in host.AddressList)
            {
                if (ip.AddressFamily == AddressFamily.InterNetwork)
                {
                    ipaddress.Add(new { IP = ip.ToString() });
                }
            }
            return ipaddress;
        }
        public class DownLoad : BeetleX.FastHttpApi.IResult
        {
            public DownLoad(string text)
            {
                Text = text;
            }

            public string Text { get; set; }

            public IHeaderItem ContentType => ContentTypes.OCTET_STREAM;

            public int Length { get; set; }

            public bool HasBody => true;

            public void Setting(HttpResponse response)
            {
                response.Header.Add("Content-Disposition", "attachment;filename=TcpBanchmarks.json");
            }

            public void Write(PipeStream stream, HttpResponse response)
            {
                stream.Write(Text);
            }
        }
    }
}

File: src/views/js/Controller.js
function Banchmark() {
    this.url_AddServer = '/AddServer';
    this.url_DelServer = '/DelServer';
    this.url_ListServer = '/ListServer';
    this.url_NewMessage = '/NewMessage';
    this.url_SaveMessage = '/SaveMessage';
    this.url_DeleteMessage = '/DeleteMessage';
    this.url_DeleteCategoryMessages = '/DeleteCategoryMessages';
    this.url_ListMessages = '/ListMessages';
    this.url_Test = '/Test';
    this.url_Download = '/Download';
    this.url_Upload = '/Upload';
    this.url_ListMessageCategories = '/ListMessageCategories';
    this.url_GetRunerDetail = '/GetRunerDetail';
    this.url_GetFileID = '/GetFileID';
    this.url_Stop = '/Stop';
    this.url_Run = '/Run';
    this.url_GetLocalIPAddress = '/GetLocalIPAddress';
}
/**
* 'AddServer(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.AddServer = function (host, useHttp) {
    return api(this.url_AddServer, { host: host }, useHttp);
}
/**
* 'DelServer(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.DelServer = function (host, useHttp) {
    return api(this.url_DelServer, { host: host }, useHttp);
}
/**
* 'ListServer(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.ListServer = function (useHttp) {
    return api(this.url_ListServer, {}, useHttp);
}
/**
* 'NewMessage(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.NewMessage = function (useHttp) {
    return api(this.url_NewMessage, {}, useHttp);
}
/**
* 'SaveMessage(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.SaveMessage = function (msg, useHttp) {
    return api(this.url_SaveMessage, { msg: msg }, useHttp, true);
}
/**
* 'DeleteMessage(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.DeleteMessage = function (id, useHttp) {
    return api(this.url_DeleteMessage, { id: id }, useHttp);
}
/**
* 'DeleteCategoryMessages(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.DeleteCategoryMessages = function (category, useHttp) {
    return api(this.url_DeleteCategoryMessages, { category: category }, useHttp);
}
/**
* 'ListMessages(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.ListMessages = function (category, useHttp) {
    return api(this.url_ListMessages, { category: category }, useHttp);
}
/**
* 'Test(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.Test = function (message, host, useHttp) {
    return api(this.url_Test, { message: message, host: host }, useHttp);
}
/**
* 'Download(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.Download = function (useHttp) {
    return api(this.url_Download, {}, useHttp);
}
/**
* 'Upload(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.Upload = function (name, completed, data, useHttp) {
    return api(this.url_Upload, { name: name, completed: completed, data: data }, useHttp);
}
/**
* 'ListMessageCategories(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.ListMessageCategories = function (useHttp) {
    return api(this.url_ListMessageCategories, {}, useHttp);
}
/**
* 'GetRunerDetail(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.GetRunerDetail = function (useHttp) {
    return api(this.url_GetRunerDetail, {}, useHttp);
}
/**
* 'GetFileID(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.GetFileID = function (useHttp) {
    return api(this.url_GetFileID, {}, useHttp);
}
/**
* 'Stop(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.Stop = function (useHttp) {
    return api(this.url_Stop, {}, useHttp);
}
/**
* 'Run(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.Run = function (server, ipaddress, messages, setting, useHttp) {
    return api(this.url_Run, { server: server, ipaddress: ipaddress, messages: messages, setting: setting }, useHttp, true);
}
/**
* 'GetLocalIPAddress(params).execute(function(result){});'
* 'FastHttpApi javascript api Generator Copyright © henryfan 2018 email:henryfan@msn.com
* 'https://github.com/IKende/FastHttpApi
**/
Banchmark.prototype.GetLocalIPAddress = function (useHttp) {
    return api(this.url_GetLocalIPAddress, {}, useHttp);
}
